lib/UART/src/UART.cpp

```diff
@@ -0,0 +1,217 @@
+#include "UART.h"
+#define BAUD 9600
+#include <avr/io.h>
+#include <avr/interrupt.h>
+#include <util/setbaud.h>
+#include <stdarg.h>
+#include <util.h>
+#include "uint8-queue.h"
+
+void init_UART(void)
+{
+    //Set baud rate
+    UBRR0H = UBRRH_VALUE;
+    UBRR0L = UBRRL_VALUE;
+
+    #if USE_2X
+        UCSR0A |= (1 << U2X0);
+    #else
+        UCSR0A &= ~(1 << U2X0);
+    #endif
+
+    cli();
+    UCSR0B = (1 << RXEN0) | (1 << TXEN0) | (1 << RXCIE0);   // Enable receiver, transmitter and receive interrupt
+    UCSR0C = (1 << UCSZ01) | (1 << UCSZ00); // Set frame format: 8 data bits, 1 stop bit, no parity
+    sei();
+}
+
+uint8_queue_t tx_queue;
+volatile const char* tx_string;
+volatile uint8_t tx_index;
+
+void UART_enable_transmit_interrupt()
+{
+    UCSR0B |= bit(UDRIE0);
+}
+
+void UART_desable_transmit_interrupt()
+{
+    UCSR0B &= ~(1 << UDRIE0);
+}
+
+void UART_print_char(const char c)
+{
+    char tx_char = c;
+    add_to_queue(&tx_queue, (uint8_t)tx_char);
+    UART_enable_transmit_interrupt();
+}
+
+void UART_print_string(const char* str)
+{
+    tx_string = str;
+    tx_index = 0;
+}
+
+void service_transmit_UART(void)
+{
+    if (tx_string == nullptr)
+    {
+        return;
+    }
+
+    if (queue_is_full(&tx_queue))
+    {
+        //ERROR: Buffer overflow
+        return;
+    }
+    else
+    {
+        UART_print_char(tx_string[tx_index++]);
+
+        if(tx_string[tx_index] == '\0')
+        {
+            tx_index = 0;
+            tx_string = nullptr;
+            return;
+        }
+    }
```

Comment:
This `else` line is redundant. You can just remove it and the surrounding curly brackets.

```suggestion
    UART_print_char(tx_string[tx_index++]);
    
    if(tx_string[tx_index] == '\0')
    {
        tx_index = 0;
        tx_string = nullptr;
        return;
    }
```